File: Completed-Algorithms/Leetcode_210_CourseSchedule2/dfs.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function(numCourses, pre) {
    // create adjacency list
    const adj = {}
    for (let i = 0; i < numCourses; i++) {
        adj[i] = []
    }
    for (let i = 0 ; i < pre.length; i++) {
        const [v1,v2] = pre[i];
        adj[v1].push(v2)
    }
    const vis = new Array(numCourses).fill(0)
    const results = []

    const dfs = n => {
        vis[n] = 2;
        for (let i = 0; i < adj[n].length; i++) {
            if(vis[adj[n][i]] === 0) {
                if(dfs(adj[n][i])) {
                    return true
                }
            }
            else if (vis[adj[n][i]] === 2) {
                return true
            }
        }
        vis[n] = 1
        results.push(n)
        return false
    }

    for(let i = 0 ; i < numCourses; i++) {
        if(vis[i] === 0) {
            if(dfs(i)) {
                return []
            }
        }
    }
    return results

};

File: Completed-Algorithms/Leetcode_210_CourseSchedule2/dfs.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = {i: [] for i in range(numCourses)}
        for a, b in prerequisites:
            graph[b].append(a)
        visited = set()
        visiting = set()
        result = []

        def dfs(node):
            if node in visiting:
                return False
            if node in visited:
                return True
            visiting.add(node)
            for nei in graph[node]:
                if not dfs(nei):
                    return False
            visiting.remove(node)
            visited.add(node)
            result.append(node)
            return True
        for node in range(numCourses):
            if not dfs(node):
                return []
        result.reverse()
        return result
    
        
